sim/firesim-lib/src/main/scala/bridges/TestFinisherBridge.scala

```diff
@@ -0,0 +1,97 @@
+
+package firesim.bridges
+
+import chisel3._
+import chisel3.util._
+import freechips.rocketchip.config.Parameters
+import midas.widgets._
+import midas.widgets.PeekPokeBridge
+
+case class CodeElaborator(isErr: Boolean, message : String)
+
+case class TestIndicatorParams(
+  nConcurrentTests: Int     = 1,
+  codeMessage: Seq[CodeElaborator]
+) {
+  def codeBits: Int    = 16
+}                 
+
+
+class TestFinisherPortIO(val params: TestIndicatorParams) extends Bundle {
+  //Error code in case program terminates unsuccessfully
+  val errCode = UInt((params.codeBits).W)
+  val valid = Bool() 
+}
+
+class TestFinisherBridgeTargetIO(params: TestIndicatorParams) extends Bundle {
+  val clock = Input(Clock())
+  val testDone = Input(new TestFinisherPortIO(params))
+}
+
+class TestFinisherBridgeHostIO(params: TestIndicatorParams)
+                              (private val targetIO : TestFinisherBridgeTargetIO = new TestFinisherBridgeTargetIO(params)) 
+                               extends Bundle with ChannelizedHostPortIO{
+  def targetClockRef = targetIO.clock 
+  //It is just inputs from the target to indicate test completion
+  //The HostPort annotation would not work with aggregates so creating individual channels.
+  val errCodeH = InputChannel(targetIO.testDone.errCode)
+  val validH = InputChannel(targetIO.testDone.valid)
+}
+
+class TestFinisherBridge(params: TestIndicatorParams)(implicit p: Parameters) extends BlackBox
+    with Bridge[TestFinisherBridgeHostIO, TestFinisherBridgeModule] {
+  val io = IO(new TestFinisherBridgeTargetIO(params))
+  val bridgeIO = new TestFinisherBridgeHostIO(params)(io)
+
+  val constructorArg = Some(params)
+
+  generateAnnotations()
+}
+
+object TestFinisherBridge {
+  def apply(clock: Clock, finisher: TestFinisherPortIO)(implicit p: Parameters): TestFinisherBridge = {
+    val tfm = Module(new TestFinisherBridge(finisher.params))
+    tfm.io.testDone := finisher
+    tfm.io.clock := clock 
+    tfm  
+  }
+}
+
+class TestFinisherBridgeModule(params: TestIndicatorParams)(implicit p: Parameters) extends BridgeModule[TestFinisherBridgeHostIO]()(p) {
+  lazy val module = new BridgeModuleImp(this) {
+    // This creates the interfaces for all of the host-side transport
+    // AXI4-lite for the simulation control bus, =
+    // AXI4 for DMA
+    val io = IO(new WidgetIO())
+
+
+
+    // This creates the host-side interface of your TargetIO
+    val hPort = IO(new TestFinisherBridgeHostIO(params)())
+    
+    val statusDone = RegInit(Bool(), false.B)
+    
+    statusDone := hPort.validH.bits
+    val errCode = RegInit(0.U((params.codeBits).W))
+    when(statusDone) {
+      errCode := hPort.errCodeH.bits
+    }
+    //Condition needed for to signal HostPort Channel can send the messages.
+    //So host is ready until the status is Done.
+    hPort.errCodeH.ready := !statusDone
```

Comment:
So i made this harder for you than it needed to be by asking you to use ChannelizedHostPortIO (since i forgot it's not possible to mark aggregates as input channels). However, you should dequeue code and valid bits carefully to handle potential decoupling between them (e.g., you could do it concurrently). You can look at other bridges for examples of this, but the simplest way to do that is as follows: 

```
val helper = DecoupledHelper(hPort.errCodeH.valid, hPort.validH.valid,  hPort.validH.bits)
hPort.errCodeH.ready := helper.fire(hPort.errorCodeH.valid)
hPort.validH.ready := helper.fire(hPort.validH.valid)
```

If you want to be more fastidious you can latch the errorcode and valid first (making sure to handle potential decoupling between the channels) , and then use those to drive each channel's ready, instead of combinationally coupling the valid from opposite channels back to ready. This is closer to what you already have.

Reply:
[https://github.com/firesim/firesim/blob/2126f267f0b61b05d86d349675674ef391a6e572/sim/firesim-lib/src/main/scala/bridges/TestFinisherBridge.scala#L70-L91](url)
I decoupled it by latching.